Frameworks/Foundation/NSUndoManager.mm

```diff
@@ -38,67 +102,159 @@ - (id)prepareWithInvocationTarget:(id)target {
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)undo {
-    UNIMPLEMENTED();
+    _groupingLevel = [_undoStack count];
+    if (_groupingLevel == 1) {
+        [self endUndoGrouping];
+    }
+
+    [self undoNestedGroup];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)undoNestedGroup {
-    UNIMPLEMENTED();
+    _groupingLevel = [_undoStack count];
+
+    if (![self canUndo]) {
+        return;
+    }
+
+    if (_groupingLevel == 0) {
+        [NSException raise:NSInternalInconsistencyException format:@"No top level group to undo."];
+    }
+
+    _undoing = YES;
+
+    _NSUndoGroup* undoGroup = [_undoStack peek];
+
+    // Record redo operation as a single redo group
+    [self _beginRedoGrouping];
+    [undoGroup undo];
+    [self _endRedoGrouping];
+    [_undoStack pop];
+
+    _groupingLevel = [_undoStack count];
+    ;
+
+    _undoing = NO;
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)redo {
-    UNIMPLEMENTED();
+    _groupingLevel = [_redoStack count];
+    if (![self canRedo]) {
+        return;
+    }
+
+    if (_undoing) {
+        [NSException raise:NSInternalInconsistencyException format:@"Cannot invoke redo while undoing is active."];
```

Comment:
>Cannot invoke redo while undoing is active. [](start = 69, length = 43)

Phrase exceptions more as a statement of fact; "-redo called during an undo operation." or something.